src/vm/functions/mod.rs

```diff
@@ -309,32 +307,39 @@ fn special_let(args: &[SymbolicExpression], env: &mut Environment, context: &Loc
     let bindings = args[0].match_list()
         .ok_or(CheckErrors::BadLetSyntax)?;
 
-    let mut binding_results = parse_eval_bindings(bindings, env, context)?;
-
-    runtime_cost!(cost_functions::LET, env, binding_results.len())?;
+    runtime_cost!(cost_functions::LET, env, bindings.len())?;
 
     // create a new context.
     let mut inner_context = context.extend()?;
 
-    for (binding_name, binding_value) in binding_results.drain(..) {
-        if is_reserved(&binding_name) ||
-           env.contract_context.lookup_function(&binding_name).is_some() ||
-           inner_context.lookup_variable(&binding_name).is_some() {
-            return Err(CheckErrors::NameAlreadyUsed(binding_name.into()).into())
-        }
-        inner_context.variables.insert(binding_name, binding_value);
-    }
+    let mut memory_use = 0;
 
-    // evaluate the let-bodies
+    finally_drop_memory!( env, memory_use; {
+        handle_binding_list::<_, Error>(bindings, |binding_name, var_sexp| {
+            if is_reserved(binding_name) ||
+                env.contract_context.lookup_function(binding_name).is_some() ||
+                inner_context.lookup_variable(binding_name).is_some() {
+                    return Err(CheckErrors::NameAlreadyUsed(binding_name.clone().into()).into())
+                }
 
-    let mut last_result = None;
-    for body in args[1..].iter() {
-        let body_result = eval(&body, env, &inner_context)?;
-        last_result.replace(body_result);
-    }
+            let binding_value = eval(var_sexp, env, context)?;
```

Comment:
Just confirming -- `eval()` will fail if the memory budget is exceeded while evaluating `var_sexp`, right?  I.e. the memory use calculation below will only occur if the code is still under-budget?  The same pattern occurs in a few other places.

Reply:
Yep

Reply:
Cool, thanks